cmd/gce/gce.go

```diff
@@ -29,140 +24,30 @@ func init() {
-// loading metadata from GCE's instance metadata.
-// https://cloud.google.com/compute/docs/storing-retrieving-metadata
-func WithGCEConfig(conf config.Config, metaURL string) (config.Config, error) {
-	log.Info("Discovering GCE environment")
 
-	// Check that this is a GCE VM environment.
-	// If not, fail.
-	meta, err := getMetadata(metaURL)
-	if err != nil {
-		log.Error("Error getting GCE metadata", err)
-		return conf, fmt.Errorf("can't find GCE metadata. This command requires a GCE environment")
-	}
-
-	log.Info("Loaded GCE metadata")
-	log.Debug("GCE metadata", meta)
+		// Check that this is a GCE VM environment.
+		// If not, fail.
+		meta, merr := gce.LoadMetadata()
+		if merr != nil {
+			log.Error("Error getting GCE metadata", merr)
+			return fmt.Errorf("can't find GCE metadata. This command requires a GCE environment")
+		}
 
-	conf.Scheduler = "gce"
-	defaultHostName := conf.HostName
+		log.Info("Loaded GCE metadata")
+		log.Debug("GCE metadata", meta)
 
-	// Load full config doc from metadata
-	if meta.Instance.Attributes.FunnelConfig != "" {
-		mconf := config.Config{}
 		var err error
-		b := []byte(meta.Instance.Attributes.FunnelConfig)
-		err = config.Parse(b, &mconf)
-		if err != nil {
-			return conf, err
-		}
-		err = mergo.MergeWithOverwrite(&conf, mconf)
+		conf, err = gce.WithMetadataConfig(conf, meta)
 		if err != nil {
-			return conf, err
-		}
-	}
-
-	// Is this a worker node? If so, inherit the worker ID from the GCE instance name.
-	if meta.Instance.Attributes.FunnelWorkerServerAddress != "" {
-		if conf.Worker.ID == "" {
-			conf.Worker.ID = meta.Instance.Name
-		}
-		conf.Worker.ServerAddress = meta.Instance.Attributes.FunnelWorkerServerAddress
-	}
-	if meta.Project.ProjectID != "" {
-		conf.Backends.GCE.Project = meta.Project.ProjectID
-	}
-	// TODO need to parse zone?
-	if meta.Instance.Zone != "" {
-		zone := meta.Instance.Zone
-		// Parse zone out of metadata format
-		// e.g. "projects/1234/zones/us-west1-b" => "us-west1-b"
-		idx := strings.LastIndex(zone, "/")
-		if idx != -1 {
-			zone = zone[idx+1:]
-		}
-		conf.Backends.GCE.Zone = zone
-	}
-
-	conf.Worker.Metadata["gce"] = "yes"
-
-	// If the configuration contains a worker ID, assume that a worker
-	// process should be started (instead of a server).
-	if conf.Worker.ID != "" {
-		if conf.Worker.ServerAddress == "" {
-			log.Error("Empty server address while starting worker")
-			return conf, fmt.Errorf("Empty server address while starting worker")
+			return err
 		}
-		conf.Worker.Storage.GS = append(conf.Worker.Storage.GS,
-			config.GSStorage{FromEnv: true})
-	}
-
-	// Auto detect the server's host name when it's not already set.
-	// This makes server deployment and configuration a bit easier.
-	// TODO will this work across zones?
-	if conf.HostName == defaultHostName && meta.Instance.Hostname != "" {
-		conf.HostName = meta.Instance.Hostname
-	}
 
-	return conf, nil
-}
+		logger.Configure(conf.Logger)
```

Comment:
This should not be called until you know if your running a worker or server process. That way you know which Logger to load from the config.